ARROW-3489: [Gandiva][C++] Added support for IN expressions

praveenbingo opened the pull request:
Gandiva now supports processing in expressions.

praveenbingo commented on cpp/src/gandiva/tree_expr_builder.h:
```diff
@@ -85,6 +86,25 @@ class TreeExprBuilder {
                                     const FieldVector& in_fields);
+
+  /// \brief creates an in expression
+  static NodePtr MakeInExpressionInt(NodePtr function,
```
yup tested for same but forgot to change the name :) fixed now.

pravindra commented:
We have an alternative to in-expression : a in (c1, c2, c3, .., cn) can also be written as (a == c1 || a == c2 || .. || a == cn). Can you please add a test in the microbenchmarks to compare the perf of the two ? 
1. we'll know the threshold n at which "in expression" outperforms the ||
2. it'll test the efficiency of the boost variant (i'm sceptical about it's perf in this use case)

praveenbingo commented on cpp/src/gandiva/function_holder_stubs.cc:
```diff
@@ -36,3 +37,28 @@ extern "C" int64_t to_date_utf8_utf8_int32(int64_t ptr, const char* data, int da
                    out_valid);
 }
+
+extern "C" bool in_expr_lookup_int32(int64_t ptr, int32_t value, bool in_validity) {
```
Good idead, but we are getting rid of the helpers library in another PR, the symbols.map will go away after that for the helpers. 
Leaving it as is to be consistent with other precompiled functions.

praveenbingo commented on cpp/src/gandiva/tests/in_expr_test.cc:
```diff
@@ -0,0 +1,140 @@
+
+  std::shared_ptr<Filter> filter;
+  Status status = Filter::Make(schema, condition, &filter);
+  EXPECT_TRUE(status.ok());
```
Yup there is another Jira for removing redundancies between Arrow and Gandiva like status, assert etc.

praveenbingo commented on cpp/src/gandiva/tests/in_expr_test.cc:
```diff
@@ -0,0 +1,140 @@
+  auto node_f1 = TreeExprBuilder::MakeField(field1);
+  auto sum_func =
+      TreeExprBuilder::MakeFunction("add", {node_f0, node_f1}, arrow::int32());
+  std::unordered_set<int32_t> in_constants;
```
Thanks :) this is simple.

praveenbingo commented:
Travis has passed, unrelated failure in Rust due to throttling by Travis.

pravindra commented on cpp/src/gandiva/expr_decomposer.cc:
```diff
@@ -173,6 +175,23 @@ Status ExprDecomposer::Visit(const BooleanNode& node) {
+    auto status = node.eval_expr()->Accept(*this);                            \
+    GANDIVA_RETURN_NOT_OK(status);                                            \
+    args.push_back(result());                                                 \
+    /* In always outpus valid results, so no validity dex */                  \
```
typo : outpus -> outputs

pravindra commented on cpp/src/gandiva/gandiva_aliases.h:
```diff
@@ -19,8 +19,13 @@
+#include <unordered_set>
 #include <vector>
 
+#include <boost/functional/hash.hpp>
```
why here ? this file is included in any places - can we move it to only the files where it's required ?

pravindra commented on cpp/src/gandiva/gdv_function_stubs.cc:
```diff
@@ -78,6 +105,34 @@ void ExportedStubFunctions::AddMappings(Engine& engine) const {
+
+  // gdv_fn_in_expr_lookup_int64
+  args = {types.i64_type(),  // int64_t in holder ptr
+          types.i64_type(),  // int value
```
int64 value ?

pravindra commented on cpp/src/gandiva/in_holder.h:
```diff
@@ -0,0 +1,49 @@
+    }
+  }
+
+  bool HasValue(Type value) const { return values_.count(value) == 1; }
```
will (find(value) != values_.end()) be faster ?

praveenbingo replied:
its the same from our benchmarks..since set does not allow for duplicates..
was originally using find switched to this to see if its faster but its the same..

pravindra commented on cpp/src/gandiva/node.h:
```diff
@@ -19,8 +19,11 @@
+#include <unordered_set>
 #include <vector>
 
+#include <boost/lexical_cast.hpp>
```
is this used ?

pravindra commented on cpp/src/gandiva/symbols-helpers.map:
```diff
@@ -15,6 +15,6 @@
 # under the License.
 {
-  global: extern "C++" { gandiva*; like*;to_date*};
+  global: extern "C++" { gandiva*; like*;to_date*; in_expr*};
```
should we remove the file ?

praveenbingo commented:
@pitrou I've addressed all of the comments and the PR is good to go. 

I am getting unrelated python code style errors in CI. Do you want to fix them as part of this PR?

pitrou commented:
The Python errors are ARROW-3606. To work around them (so that we have a CI run) please apply https://github.com/apache/arrow/pull/2812/commits/1adbe8a2bd81665ef5f182ad6a99f79299509293 (and undo your Python changes) :-)

praveenbingo commented:
@pitrou can you please merge if this looks ok to you..got a clean run on CI for everything except R(which is getting throttled)..

pitrou commented on cpp/src/gandiva/expr_validator.cc:
```diff
@@ -154,4 +154,67 @@ Status ExprValidator::Visit(const BooleanNode& node) {
+ * 1. Non empty list of constants to search in.
+ * 2. Expression returns of the same type as the constants.
+ */
+Status ExprValidator::Visit(const InExpressionNode<int32_t>& node) {
```
The code duplication in these three overloads looks gratuitous to me. Can't you factor them out?

praveenbingo replied:
Yes I should!

pitrou commented on cpp/src/gandiva/dex.h:
```diff
@@ -272,6 +275,73 @@ class BooleanOrDex : public BooleanDex {
+};
+
+template <>
+class InExprDex<int32_t> : public InExprDexBase<int32_t> {
```
The three specializations look almost the same (except the name of the runtime function). Can't you use common code for them?

praveenbingo replied:
yes i should! moved everything other than runtime function to base class..

pitrou commented on cpp/src/gandiva/gdv_function_stubs.cc:
```diff
@@ -78,6 +105,34 @@ void ExportedStubFunctions::AddMappings(Engine& engine) const {
                                  reinterpret_cast<void*>(gdv_fn_to_date_utf8_utf8_int32));
+
+  // gdv_fn_in_expr_lookup_int32
+  args = {types.i64_type(),  // int64_t in holder ptr
```
I'm curious: is it possible to declare those signatures in the same file as the function they describe? It would probably makes things easier to follow and maintain.

And/or perhaps have some kind of declarative registry mechanism? I expect more runtime helpers will be needed along the way...

praveenbingo replied:
Agreed we will definitely have more helpers coming soon. We can move these to the individual helper files.

By declarative do you mean some kind of config that automatically registers a holder to this list of exported functions?

(also can we address this in a follow up PR, i will update 3489 that this needs to be fixed)

pitrou replied:
I mean some kind of macro or construct that would allow those to be top-level declarations somewhere. Agreed about addressing that in a followup PR, though.

pitrou commented on cpp/src/gandiva/llvm_generator.cc:
```diff
@@ -832,6 +832,46 @@ void LLVMGenerator::Visitor::Visit(const BooleanOrDex& dex) {
   result_.reset(new LValue(result_value));
 }
 
+#define MAKE_VISIT_IN(ctype)                                                           \
```
I'm not sure why you define a macro for this rather than a private templatized helper func. Long macros are tedious to maintain and debug.

praveenbingo replied:
i plead ignorance :( did not know functions could be templatized too..
changed it to a template helper function.

pitrou commented on cpp/src/gandiva/tests/generate_data.h:
```diff
@@ -46,7 +46,11 @@ class BoundedInt32DataGenerator : public Int32DataGenerator {
 
-  int32_t GenerateData() { return (rand_r(&seed_) % upperBound_); }
+  int32_t GenerateData() {
+    int32_t value = (rand_r(&seed_) % upperBound_);
```
Why don't you use [C++ random number generation](https://en.cppreference.com/w/cpp/numeric/random) in this header?

praveenbingo replied:
sure changed.

curious is c++ generator better in some way? (could not find in a quick google search).

pitrou replied:
It probably depends on the libc. It provides many more options, though.

pitrou commented on cpp/src/gandiva/tests/micro_benchmarks.cc:
```diff
@@ -213,4 +214,82 @@ TEST_F(TestBenchmarks, TimedTestFilterLike) {
+// following two tests are for benchmark optimization of
+// in expr. will be used in follow-up PRs to optimize in expr.
+
+TEST_F(TestBenchmarks, TimedTestMultiOr) {
```
Wow, we shouldn't put benchmarks in unit tests. We use separate gbenchmark-based executables for that. You can look for examples in the Arrow source tree.

praveenbingo replied:
agreed, we've thinking about it..created ARROW-3609 for the same.

pitrou commented on cpp/src/gandiva/tree_expr_builder.h:
```diff
@@ -85,6 +86,25 @@ class TreeExprBuilder {
+  /// \brief creates an in expression
+  static NodePtr MakeInExpressionInt(NodePtr node,
+                                     const std::unordered_set<int32_t>& constants);
+  static NodePtr MakeInExpressionLong(NodePtr node,
```
I'm not sure why those are suffixed `Int`/`Long` rather than `Int32`/`Int64`? A Java reference perhaps? :-)

praveenbingo replied:
ha yes..old habits :) will change them

praveenbingo commented:
@pitrou - thanks for the thorough review :) can you please merge if this looks ok now?

praveenbingo commented:
thanks @pitrou
